base/strings/lazy.jl

```diff
@@ -63,14 +72,15 @@ macro lazy_str(text)
 end
 
 function String(l::LazyString)
-    if !isdefined(l, :str)
-        l.str = sprint() do io
-            for p in l.parts
-                print(io, p)
-            end
+    old = getfield(l, :str, :acquire)
+    old === nothing || return old
+    str = sprint() do io
+        for p in l.parts
+            print(io, p)
         end
     end
-    return l.str
+    old = swapfield!(l, :str, str, :acquire_release)
```

Comment:
You appear to have the `@atomic` macros, so does this work?
```suggestion
    old = @atomicswap :acquire_release l.str = str
```

Reply:
> Or is it better to define some array methods to earlier?

Maybe this is simple enough: b8dd0808532b03aa2f9f29897020b7271f81999e